kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/ApplicationStream.java

```diff
@@ -0,0 +1,248 @@
+    private final BigQueryWriteClient client;
+    /**
+     * Number of times append is called
+     */
+    private final AtomicInteger appendCallsCount;
+    /**
+     * Number of append requests completed successfully. This can never be greater than appendCallsCount
+     */
+    private final AtomicInteger completedCallsCount;
+
+    /**
+     * This is called by builder to guarantee sequence.
+     */
+    private final AtomicInteger maxCallsCount;
+
+    public ApplicationStream(String tableName, BigQueryWriteClient client) throws Exception {
+        this.client = client;
+        this.tableName = tableName;
+        this.offsetInformation = new HashMap<>();
+        this.appendCallsCount = new AtomicInteger();
+        this.maxCallsCount = new AtomicInteger();
+        this.completedCallsCount = new AtomicInteger();
+        generateStream();
+        currentState = StreamState.CREATED;
+    }
+
+    public Map<TopicPartition, OffsetAndMetadata> getOffsetInformation() {
+        return offsetInformation;
+    }
+
+    private void generateStream() throws Descriptors.DescriptorValidationException, IOException, InterruptedException {
+        this.stream = client.createWriteStream(
+                tableName, WriteStream.newBuilder().setType(WriteStream.Type.PENDING).build());
+        this.jsonWriter = JsonStreamWriter.newBuilder(stream.getName(), client).build();
+    }
+
+    public void closeStream() {
+        if (!this.jsonWriter.isClosed()) {
+            this.jsonWriter.close();
+            logger.info("JSON Writer for stream {} closed", getStreamName());
+        }
+    }
+
+    public String getStreamName() {
+        return this.stream.getName();
+    }
+
+    /**
+     * Increases the Append call count and returns the updated value
+     */
+    public void increaseAppendCallCount() {
+        this.appendCallsCount.incrementAndGet();
+    }
+
+    /**
+     * Increases the Max call count by 1. This tells the total expected calls which would be made to append method.
+     * Returns the updated value
+     */
+    public int increaseMaxCallsCount() {
+        int count = this.maxCallsCount.incrementAndGet();
+        if (currentState == StreamState.CREATED) {
+            currentState = StreamState.APPEND;
+        }
+        return count;
+    }
+
+    /**
+     * Increases the count of Append calls which are completed.
+     * Returns the updated value
+     */
+    public void increaseCompletedCallsCount() {
+        this.completedCallsCount.incrementAndGet();
+    }
+
+    /**
+     * Stream can be closed for writing (not appending new data) only if its current state is different from created
+     * A stream with CREATED state tells the stream has not been used for writing anything and would result in resource
+     * wastage we create new without using the existing one
+     *
+     * @return
+     */
+    public boolean canBeMovedToNonActive() {
```

Comment:
Inactive is a different state actually which means the stream has completed its lifecycle. Here non-active means, the stream already has data assigned to it, can be pushed back in list so that any new data gets assigned to a new stream. 
This stream would be later eligible for commit. If we keep writing everything in one stream we would not be able to commit it. A committed stream cannot be re-used.

Reply:
How about `canTransitionToNonActive` ?

Reply:
Sure ... we can go with `canTransitionToNonActive`